Justfile

```diff
@@ -91,20 +91,30 @@ upgrade env package="": _virtualenv
     FORCE=true {{ just_executable() }} requirements-{{ env }} $opts
 
 
+black *args=".": devenv
+    $BIN/black --check {{ args }}
+
+
+isort *args=".": devenv
+    $BIN/isort --check-only --diff {{ args }}
+
+
+flake8 *args="": devenv
+    $BIN/flake8 {{ args }}
+
+
+pyupgrade *args="$(find databuilder -name '*.py' -type f) $(find tests -not -path 'tests/acceptance/external_studies/*' -name '*.py' -type f)": devenv
+    $BIN/pyupgrade --py311-plus --keep-percent-format {{ args }}
+
+
 # runs the format (black), sort (isort) and lint (flake8) checks but does not change any files
-check: devenv
-    $BIN/black --check .
-    $BIN/isort --check-only --diff .
-    $BIN/flake8
-    $BIN/pyupgrade --py39-plus --keep-percent-format \
-        $(find databuilder -name "*.py" -type f) \
-        $(find tests -not -path 'tests/acceptance/external_studies/*' -name "*.py" -type f)
+check: black isort flake8 pyupgrade docstrings
     just docstrings
     docker pull hadolint/hadolint
     docker run --rm -i hadolint/hadolint < Dockerfile
 
 # ensure our public facing docstrings exist so we can build docs from them
-docstrings: devenv
+docstrings *args: devenv
```

Comment:
Yes, it does need it, because precommit passes the changed files along and then `just` interprets them as commands, so I get errors like:
```error: Justfile does not contain recipe `databuilder/backends/tpp.py```

Reply:
Oh yes, fun!